### ISSUE TITLE
Speed up action installation

### ISSUE THREAD

**Issue by fabiosangregorio** [zemuldo/git-crypt-unlock#1]
It currently takes ~30s on my build to install this action. Is there a way to speed this up, maybe by caching the docker build?

**Comment by fabiosangregorio** [comment]
Closing this since I'm not getting any activity.

**Comment by zemuldo** [comment]
@fabiosangregorio Sorry, I was lost on some stuff, I know about the speed of this action and I have started work out non container based version for this (trial). I will push a PR for you to checkout :-)
 have reopened this issue

**Comment by fabiosangregorio** [comment]
Awesome! Can't wait to check it out.

Since this is going to be non container based, does this mean that I will be able to commit & push files after unlocking the repo with this action? Currently issuing a `git commit -m ...` after unlocking the files results in the following git error:
```
"git-crypt" clean: 1: "git-crypt" clean: git-crypt: not found
error: external filter '"git-crypt" clean' failed 127
error: external filter '"git-crypt" clean' failed
fatal: telereddit/config/secret_dev.py: clean filter 'git-crypt' failed
```

This can be seen in this Workflow run example: https://github.com/fabiosangregorio/telereddit/runs/632947199

I think this is due to git-crypt being in a container and not visible to the workflow environment after the unlock.
I'm currently manually installing all dependencies and setting up git-crypt due to this, but I would be really happy to be going back to use this action if the issue is solved. Should I open a separate issue for this?

**Comment by zemuldo** [comment]
If I get you properly, the new version will run directly in the runtime that executes your actions so I think it should work with your scenario because the commands used by the this action will be accessible directly in your runtime. Lets check it out when done :-)

**Comment by zemuldo** [comment]
Checkout https://github.com/zemuldo/zemuldo.com/runs/654232811

Here it took 10 seconds with no cache. I will try to find out how to reduce further with cache

All you need todo is to update `zemuldo/git-crypt-unlock@master` to  `zemuldo/git-crypt-unlock@no-docker`

If this works for you I will improve it and push a PR for you. Cheers!

**Comment by zemuldo** [comment]
Note, this version only works on unix - Mac or Linux runtimes
I tried to work around windows but seems to require more work and digging.

**Comment by zemuldo** [comment]
@fabiosangregorio ☝️

**Comment by fabiosangregorio** [comment]
Wow, I just tested it on my workflows and it took 5 seconds to build up and worked perfectly. Thank you very much.

Just a note: in my test workflow, before git-crypt unlocking my files I assert that they are locked so that if for some reason they aren't, I can notice it and take security actions.

I do this with the following bash script:
```
- name: Ensure git-crypt files are locked
      run: |
        grep -qsPa "\x00GITCRYPT"  $(git crypt status | grep -v '^not' | awk '{print $2}')
```

Only then I go ahead and unlock them. Of course, since `zemuldo/git-crypt-unlock@no-docker` runs after this, I get the following error:

`git: 'crypt' is not a git command. See 'git --help'.`

(See: https://github.com/fabiosangregorio/telereddit/runs/655864365?check_suite_focus=true)

Is there a way that you can setup (install and configure) git-crypt before everything else, and git-crypt unlock when the action is run? Just like you do on the docker build.

(See: https://github.com/fabiosangregorio/telereddit/runs/628970539?check_suite_focus=true)

Again, thank you very much!

**Comment by zemuldo** [comment]
On your command should you have `git crypt` or `git-crypt`? I think you are missing the -.

**Comment by zemuldo** [comment]
If you have setup correctly then files should be decrypted after the action. Also git-crypt should be accessible all the way down after the actions.
See here https://github.com/zemuldo/zemuldo.com/runs/656215004?check_suite_focus=true#step:5:13

**Comment by zemuldo** [comment]
Also you can try `git-crypt status | grep -v '^not' | awk '{print $2}'` and see if solves the error `git: 'crypt' is not a git command. See 'git --help'.`

**Comment by zemuldo** [comment]
This run https://github.com/fabiosangregorio/telereddit/runs/655864365?check_suite_focus=true does not appear to have the git-crypt-unlock action fired 🤔

**Comment by fabiosangregorio** [comment]
I might not have expressed myself well.

`git crypt status` is a valid command (and is equivalent to `git-crypt status`) and returns the encryption status for every file in the repo. E.g.:

```bash
>> git crypt status
not encrypted: setup.cfg
not encrypted: telereddit/config/secret.py
    encrypted: telereddit/config/secret_dev.py
not encrypted: telereddit/config/secret_generic.py
    encrypted: telereddit/config/secret_github.py
    encrypted: telereddit/config/secret_prod.py
not encrypted: telereddit/exceptions.py
not encrypted: telereddit/helpers.py
...
```

[This run](https://github.com/fabiosangregorio/telereddit/runs/655864365?check_suite_focus=true) (in which the git-crypt-unlock action does not fire) is associated with the following [Github workflow file](https://github.com/fabiosangregorio/telereddit/actions/runs/98995851/workflow):

```yml
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master

    - name: Ensure git-crypt files are locked
      run: |
        grep -qsPa "\x00GITCRYPT"  $(git crypt status | grep -v '^not' | awk '{print $2}')

    - name: Unlock git-crypt files
      uses: zemuldo/git-crypt-unlock@no-docker
      env: 
        GPG_PRIVATE_KEY: ${{ secrets.GPG_PRIVATE_KEY }}
        GPG_KEY_GRIP: ${{ secrets.GPG_KEY_GRIP }}
        GPG_KEY_PASS: ${{ secrets.GPG_KEY_PASS }} 
```

Therefore, `Ensure git-crypt files are locked` is executed before `Unlock git-crypt files` and thus cannot find git-crypt, since it is installed after that step.

On the other hand, in the docker version of your action (e.g. [this run](https://github.com/fabiosangregorio/telereddit/runs/628970539?check_suite_focus=true)) which is associated with [this workflow file](https://github.com/fabiosangregorio/telereddit/actions/runs/90968632/workflow):
```yml
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master

    - name: Unlock git-crypt files
      uses: zemuldo/git-crypt-unlock@master
      env: 
        GPG_PRIVATE_KEY: ${{ secrets.GPG_PRIVATE_KEY }}
        GPG_KEY_GRIP: ${{ secrets.GPG_KEY_GRIP }}
        GPG_KEY_PASS: ${{ secrets.GPG_KEY_PASS }}
```
we can see that before running any action in the run, your action gets built, and git-crypt gets installed and configured, with these steps in the run:
```
>  Set up job
>  Build zemuldo/git-crypt-unlock@master
>  Run actions/checkout@master
>  Unlock git-crypt files
```

What I'm trying to say is: is there a way that you could separate the build step (installation and configuration) and the unlock step (`git-crypt unlock` command), just like in the docker run, so that I would be able to check if files are locked before actually unlocking them?

Thank you for your patience.

**Comment by zemuldo** [comment]
Oh, I get you, Let me see.

**Comment by zemuldo** [comment]
What happens when you rewrite like this?
```yaml
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@master

    - name: Unlock git-crypt files
      uses: zemuldo/git-crypt-unlock@no-docker
      env: 
        GPG_PRIVATE_KEY: ${{ secrets.GPG_PRIVATE_KEY }}
        GPG_KEY_GRIP: ${{ secrets.GPG_KEY_GRIP }}
        GPG_KEY_PASS: ${{ secrets.GPG_KEY_PASS }} 
    - name: Ensure git-crypt files are locked
      run: |
        grep -qsPa "\x00GITCRYPT"  $(git crypt status | grep -v '^not' | awk '{print $2}')
```

**Comment by zemuldo** [comment]
@fabiosangregorio Could you rerun the action. I made some changes so that the action does setup before other actions start.

Like https://github.com/zemuldo/zemuldo.com/runs/656662032?check_suite_focus=true

**Comment by fabiosangregorio** [comment]
YES! This is perfect! I just tested it on two of my workflows and they were a success. Thank you very much!

Am I going to continue using the @no-docker action? Or is it going to be merged into master?

**Comment by zemuldo** [comment]
I will update and release then comment here for you to update. Then we can close this. Meanwhile, you can use `@no-docker` till then.

**Comment by zemuldo** [comment]
I have published this a pr-release of version 3 (find here 👉  [v3.0-alpha-1](https://github.com/marketplace/actions/git-crypt-unlock?version=v3.0-alpha-1)).  I also upgraded version 1 to 2 with a prebuilt image much faster, thanks to you.
I will try to support other runtimes on this non docker setup then publish as version 3

So for now use this [v3.0-alpha-1](https://github.com/marketplace/actions/git-crypt-unlock?version=v3.0-alpha-1) and close the issue. When done I will publish in version 3.0
